File: Practica1/Tarea1/cliente.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <winsock2.h>   // biblioteca de sockets para Windows (Winsock2)
#pragma comment(lib, "ws2_32.lib") // enlaza automáticamente la librería ws2_32 al compilar con MSVC o MinGW

#define PUERTO   5000   // debe coincidir con el puerto que usa el servidor Java
#define IP       "127.0.0.1" // dirección del servidor (localhost = misma máquina)
#define TAM_BUF  256    // tamaño máximo del buffer para mensajes de texto

int main() {
    WSADATA wsaData;
    WSAStartup(MAKEWORD(2, 2), &wsaData); // inicializa la librería Winsock; obligatorio antes de usar cualquier función de socket en Windows

    SOCKET sock = socket(AF_INET, SOCK_STREAM, 0); // crea un socket TCP (AF_INET = IPv4, SOCK_STREAM = orientado a conexión)
    if (sock == INVALID_SOCKET) {
        printf("Error al crear el socket.\n");
        WSACleanup();
        return 1;
    }

    struct sockaddr_in servidor; // estructura que contiene la dirección y puerto del servidor al que nos conectaremos
    servidor.sin_family      = AF_INET;        // familia de direcciones IPv4
    servidor.sin_port        = htons(PUERTO);  // convierte el puerto a orden de bytes de red (big-endian)
    servidor.sin_addr.s_addr = inet_addr(IP);  // convierte la cadena IP a formato binario de red

    if (connect(sock, (struct sockaddr*)&servidor, sizeof(servidor)) == SOCKET_ERROR) { // establece la conexión TCP con el servidor
        printf("No se pudo conectar al servidor.\n");
        closesocket(sock);
        WSACleanup();
        return 1;
    }
    printf("Conectado al servidor.\n");

    // --- ENVÍO DEL MENSAJE AL SERVIDOR ---
    char mensaje[] = "Hola\n"; // el '\n' es importante: Java usa readLine() que espera un salto de línea como delimitador
    send(sock, mensaje, strlen(mensaje), 0); // envía los bytes del mensaje por el socket al servidor
    printf("Mensaje enviado: %s", mensaje);

    // --- RECEPCIÓN DE LA RESPUESTA DEL SERVIDOR ---
    char buffer[TAM_BUF];
    memset(buffer, 0, TAM_BUF); // limpia el buffer antes de recibir para evitar basura residual en memoria
    recv(sock, buffer, TAM_BUF - 1, 0); // recibe los bytes enviados por el servidor Java (bloqueante)
    printf("Respuesta del servidor: %s\n", buffer);

    // cierre ordenado del socket y liberación de Winsock
    closesocket(sock);
    WSACleanup(); // libera todos los recursos internos que Winsock reservó con WSAStartup
    return 0;
}

File: Practica1/Tarea2/Cliente.java
import java.io.*;
import java.net.*;
import java.util.Scanner;

public class Cliente {

    public static void main(String[] args) {
        String ip    = "127.0.0.1"; // dirección del servidor (localhost = misma máquina)
        int    puerto = 5000;        // debe coincidir con el puerto del servidor C

        try {
            Socket socket = new Socket(ip, puerto); // establece la conexión TCP con el servidor C
            System.out.println("Conectado al servidor C.");

            // streams para enviar y recibir datos por el socket
            PrintWriter salida  = new PrintWriter(new OutputStreamWriter(socket.getOutputStream()), true); // envía texto al servidor (autoflush activado)
            BufferedReader entrada = new BufferedReader(new InputStreamReader(socket.getInputStream()));   // recibe texto del servidor

            Scanner teclado = new Scanner(System.in); // lee los enteros que el usuario escribe por consola

            while (true) {
                System.out.print("Ingresa un número entero (0 para salir): ");
                int numero = teclado.nextInt(); // captura el entero ingresado por el usuario

                salida.println(numero); // envía el número al servidor C como texto terminado en '\n'

                String respuesta = entrada.readLine(); // espera y lee la respuesta del servidor C (bloqueante)
                int numeroRespuesta = Integer.parseInt(respuesta.trim()); // convierte la respuesta de String a int

                System.out.println("Respuesta del servidor: " + numeroRespuesta);

                if (numero == 0) { // si el usuario ingresó 0, termina el programa
                    System.out.println("Enviaste 0. Cerrando conexión.");
                    break;
                }
            }

            // cierre ordenado de recursos
            teclado.close();
            entrada.close();
            salida.close();
            socket.close();

        } catch (IOException e) {
            System.err.println("Error en el cliente: " + e.getMessage()); // muestra el error si falla la conexión o la transmisión
        }
    }
}